tests/cache_test.py

```diff
@@ -172,3 +171,40 @@ def lambda_handler(event, context):
             return 'Hello from Lambda with secret %s' % secret_value
         
         lambda_handler(None, None)
+
+    # This isn't any different than the default behaviour, but allows
+    # us to confirm that the code calls the provided factory method.
+    def _get_client(self, session, name):
+        self.got_client = True
+        return session.client(name)
+
+    # This isn't any different than the default behaviour, but allows
+    # us to confirm that the code calls the provided factory method.
+    def _get_session(self):
+        self.got_session = True
+        return boto3.session.Session()
+
+    def test_session_factory(self):
+        # Clear the cached session and client from previous tests
+        SSMParameter.BOTO3_SESSION = None
+        SSMParameter.BOTO3_CLIENTS = {}
+
+        # Set the session and client factories
+        SSMParameter.BOTO3_SESSION_FACTORY = self._get_session
+        SSMParameter.BOTO3_CLIENT_FACTORY = self._get_client
+
+        self.got_client = False
+
+        cache = SSMParameter("my_param")
+        my_value = cache.value
+        self.assertEqual(my_value, self.PARAM_VALUE)
+
+        self.assertTrue(self.got_session)
+        self.assertTrue(self.got_client)
+
+        # Reset
```

Comment:
You could avoid this kind of resetting behaviour with `mock.patch`. Something like the following:

```python
from mock import patch
@patch('SSMParameter.BOTO3_SESSION_FACTORY', new=session_factory)
@patch('SSMParameter. BOTO3_CLIENT_FACTORY', new=client_factory)
def test_session_factory(self):
    # your tests here (without any [re-]set)
```

(patch will automatically patch the class attributes and reset them after that individual test)

Reply:
👍